File: backtracking/n_queens.js
/**
 * N-Queens Problem
 */

function isSafe(board, row, col) {
    const n = board.length;

    // Check col
    for (let i = 0; i < row; i++) {
        if (board[i][col] === 1) return false;
    }

    // Check upper left diagonal
    for (let i = row, j = col; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j] === 1) return false;
    }

    // Check upper right diagonal
    for (let i = row, j = col; i >= 0 && j < n; i--, j++) {
        if (board[i][j] === 1) return false;
    }

    return true;
}

function printBoard(board) {
    for (const row of board) {
        console.log(row.map(cell => (cell ? "Q" : ".")).join(" "));
    }
    console.log("");
}

let solutions = 0;

function solve(board, row) {
    if (row >= board.length) {
        solutions++;
        printBoard(board);
        return;
    }

    for (let col = 0; col < board.length; col++) {
        if (isSafe(board, row, col)) {
            board[row][col] = 1;
            solve(board, row + 1);
            board[row][col] = 0; // Backtrack
        }
    }
}

const n = 4;
const board = Array.from({ length: n }, () => Array(n).fill(0));
solve(board, 0);
console.log("Total solutions:", solutions);

File: backtracking/n_queens.rs
fn is_safe(board: &Vec<Vec<u8>>, row: usize, col: usize) -> bool {
    let n = board.len();

    // Check column
    for i in 0..row {
        if board[i][col] == 1 { return false; }
    }

    // Check upper left diagonal
    let (mut i, mut j) = (row as i32, col as i32);
    while i >= 0 && j >= 0 {
        if board[i as usize][j as usize] == 1 { return false; }
        i -= 1;
        j -= 1;
    }

    // Check upper right diagonal
    let (mut i, mut j) = (row as i32, col as i32);
    while i >= 0 && j < n as i32 {
        if board[i as usize][j as usize] == 1 { return false; }
        i -= 1;
        j += 1;
    }

    true
}

fn print_board(board: &Vec<Vec<u8>>) {
    for row in board {
        for cell in row {
            print!("{} ", if *cell == 1 { "Q" } else { "." });
        }
        println!();
    }
    println!();
}

fn solve(board: &mut Vec<Vec<u8>>, row: usize, solutions: &mut u32) {
    if row >= board.len() {
        *solutions += 1;
        print_board(board);
        return;
    }

    for col in 0..board.len() {
        if is_safe(board, row, col) {
            board[row][col] = 1;
            solve(board, row + 1, solutions);
            board[row][col] = 0;
        }
    }
}

fn main() {
    let n = 4;
    let mut board = vec![vec![0; n]; n];
    let mut solutions = 0;
    solve(&mut board, 0, &mut solutions);
    println!("Total solutions: {}", solutions);
}
